## Finite Field Realisation of the Classical Constants π and e, [Akhtman, Preprints 2026](https://doi.org/10.20944/preprints202607.1141.v1)

Each cell below verifies one predicate of the paper's predicate ledger (Appendix A; [finitering.space/13-epi/](https://finitering.space/13-epi/)): it runs the
blocks of [`epi.py`](https://finitering.space/src/13-epi/) whose checks cite the predicate, prints the check and lists every record that cites it with its verdict. Any cell can be run first, or *Runtime → Run all* (≈ 60 s; the C pass of block kur is the slow part).

Every cell is self-contained: its first line installs the package from the site (pip reports it already satisfied once it is there), its last runs the predicate.

In [1]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:B2 (p13009) [T] — the deciding family tow.O
# Orientation transport: under $\gen'=\gen^{u}$, $\gcd(u,\p-1)=1$, the re-derived quarter-turn is $\im'=\im$
# exactly when $u\equiv1\pmod4$ and $\im'=-\im$ (the conjugate chart) when $u\equiv3\pmod4$; there is no third
# case, and every chirality-sensitive statement holds for the other chirality under $\im\mapsto-\im$.
from frc_13_epi import predicate; predicate("13:B2")

height run: 500 shells, median H/sqrt(p) = 0.525, H<=10: 68, height-2: [13, 1933, 4177, 5857]
  [PASS] tow.E   EXACT [13:F7, 13:B3] the fixed-shell tower of e on p = 13 and 29, grades m ≤ 3: !(mp) ≡ (−1)^m, the shell reading [q_m] = e_p exact  --  18 exact micro-checks
  [PASS] tow.P   EXACT [13:G9, 13:B3] the fixed-shell tower of π on p = 13 and 29, n = p, p²: C(2n,n) ≡ 2, 16ⁿ ≡ 16, δ_p = ⟨−34⟩, the shell reading   --  20 exact micro-checks
  [PASS] tow.C   EXACT [13:C3] the Cayley quarter-turn map C(x) = (1 + ix)/(1 − ix) on F_13 and F_29: C(0) = 1, C(1) = i, and the composition  --  6 exact micro-checks
  [PASS] tow.O   EXACT [13:B2] orientation transport on F_13 and F_29, exhaustively over the units u of the exponent cycle: g' = g^u re-deriv  --  2 exact micro-checks
  [PASS] tow.H   EXACT [13:J2, 13:J3, 13:J5] the height run over all 500 shells p ≡ 1 (mod 4), p ≤ 8009, in the smallest-primitive-root frame with H minimi  --  4 exact micro-checks; median H/sqrt(p) = 0.525 (diagnosti

In [2]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:B3 (p13010) [T] — the deciding family tow.E
# The coincidence that forces the typing: at $\p=13$ the field element $6$ is at once $\eP$, $\piA$ and the
# reading of $-1/2$, while $\eR\ne\piR$; no map from a bare residue to a real number exists, and the external
# comparison takes the structural role as part of its input --- roles, not residues, carry external meaning.
# Witnessed by the two towers reading the same element $6$ as $\eP$ and as $\piA$.
from frc_13_epi import predicate; predicate("13:B3")

— epi.py:1118 (the check that decides 13:B3; family tow.E)
 1118      # 13:B3 (p13010), 13:F7 (p13033)
 1119      family("tow", "E")
 1120      for m in (1, 2, 3):
 1121          delta = centered(((-1)**m * eP) % p, p)
 1122          num = math.factorial(m*p) + delta
 1123          den = subfact(m*p)
 1124          check_tow(f"p={p} e-tower grade {m}: !(mp) == (-1)^m (mod p)", den % p == ((-1)**m) % p)
 1125          check_tow(f"p={p} e-tower grade {m}: shell reading = e_p", (num * pow(den % p, p-2, p)) % p == eP)
 1126          lo, hi = e_brackets(max(m*p//2, 12))
 1127          q = F(num, den)
 1128          # the paper's certificate: |q - e| < (mp)!/( !(mp) !(mp+1) ) + pi_F/!(mp)
 1129          width = F(math.factorial(m*p), den*subfact(m*p+1)) + F((p-1)//2, den)
 1130          check_tow(f"p={p} e-tower grade {m}: |q - e| inside certified window",
 1131                q > lo - width and q < hi + width)
  [PASS] tow.E    (deciding)       18 exact micro-checks
  [PASS] tow.P    (corro

In [3]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:C3 (p13014) [T] — the deciding family tow.C
# The Cayley quarter-turn map $C(x)=(1+\im x)/(1-\im x)$ on $\Fp$ (denominators units): $C(0)=1$, $C(1)=\im$,
# and $C(x)C(y)=C\bigl((x+y)/(1-xy)\bigr)$, the rational composition law being the half-angle form of phase
# composition; shell arithmetic only.
from frc_13_epi import predicate; predicate("13:C3")

— epi.py:1154 (the check that decides 13:C3; family tow.C)
 1154      # 13:C3 (p13014)
 1155      family("tow", "C")
 1156      cnt = 0
 1157      for x in range(p):
 1158          for y in range(p):
 1159              dx, dy = (1 - i*x) % p, (1 - i*y) % p
 1160              dxy = (1 - x*y) % p
 1161              if dx and dy and dxy:
 1162                  num_c = ((x + y) * pow(dxy, p-2, p)) % p
 1163                  dc = (1 - i*num_c) % p
 1164                  if dc:
 1165                      Cx = ((1 + i*x) * pow(dx, p-2, p)) % p
 1166                      Cy = ((1 + i*y) * pow(dy, p-2, p)) % p
 1167                      Cc = ((1 + i*num_c) * pow(dc, p-2, p)) % p
  [PASS] tow.C    (deciding)       6 exact micro-checks
13:C3: VERIFIED — 1 family record(s)


In [4]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:C5 (p13016) [T] — the deciding family pi.R1
# The joint-return tally: with $C_n=\Cb{n}$ the returning $\pm1$ histories of length $2n$ in one direction,
# the paired histories in the directions $1$ and $\im$ number $16^{n}$ and the joint returns $C_n^{2}$, and
# $\piR=\lim16^{n}/(nC_n^{2})$ by the Wallis bounds (E2).
from frc_13_epi import predicate; predicate("13:C5")

[1] RADIAL CHAINS
(a) Wallis pair: v_n < pi < w_n for n=1..300: PASS; exact width identity w-v=w/(2n+1): PASS; strict monotonicity: PASS
    example n=50: width = 3.1261e-02 ~ pi/101 = 3.1105e-02  (O(1/n) rate)
    p=13 residue line [w_n]_13, n=1..6: [4, 5, 10, 9, 6, 11] PASS
(b) Machin chain: minimal N with float(M_N) == IEEE-double(pi): N = 10
    error bound |M_N - pi| < 17/((2N+3) 5^(2N+3)) for N=2..39: PASS
    100-decimal-digit determination at N = 71
    fQ-native certified enclosure at (12,4) terms: pi in (3.1415926535897931, 3.1415926535897931): PASS
(c) precision-cost (terms n for 2^-k):
    k= 24: Wallis/Leibniz ~ 1.7e+07   Machin ~ 4   arcsin(1/2) ~ 12   (e-chain reference: 11)
    k= 53: Wallis/Leibniz ~ 9.0e+15   Machin ~ 10   arcsin(1/2) ~ 27   (e-chain reference: 18)
    k=333: Wallis/Leibniz ~ 1.7e+100   Machin ~ 71   arcsin(1/2) ~ 167   (e-chain reference: 70)

[2] WALL ARITHMETIC (mod p)
     p | window [1,2t] all legible | (2t,p) all blind | k(w_wall) | v blind at w

In [5]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:D3 (p13019) [T] — the deciding family e.R1
# Superexponential enclosure: $|\varepsilon_n-\eR|\le\eR(\eR+1)/(n+1)!<10.2/(n+1)!$ with
# $\operatorname{sign}(\varepsilon_n-\eR)=(-1)^{n+1}$, hence $\varepsilon_{2m}<\eR<\varepsilon_{2m+1}$, a
# nested sequence of $\fQ$-native intervals; the sharper constant $8$ for $2\le n\le60$.
from frc_13_epi import predicate; predicate("13:D3")

[1] METRIC EMERGENCE: the chain eps_n = n!/D_n
(a) |n!/D_n - e| < 8/(n+1)! for n=2..60 : PASS
    alternating enclosure eps_even < e < eps_odd : PASS
(b) minimal n with float(n!/D_n) == IEEE-double(e): n = 18
    n!/D_n at n=18: 6402373705728000/2355301661033953
    |eps_17 - e| = 1.096e-15,  half-ulp(e) = 2.220e-16
(c) 100-digit precision reached at n = 70; leading agreement with e: 101 chars
    e            = 2.7182818284590452353602874713526624977572470936999595749669676277240766303535475945713821785251664274274
    70!/D_70     = 2.7182818284590452353602874713526624977572470936999595749669676277240766303535475945713821785251664274188
(d) precision-cost comparison (steps n needed for 2^-k):
    k= 24:  derangement chain n =  11   |   (1+1/n)^n chain n ~ 2.28e+07
    k= 53:  derangement chain n =  18   |   (1+1/n)^n chain n ~ 1.22e+16
    k=333:  derangement chain n =  70   |   (1+1/n)^n chain n ~ 2.38e+100

[2] FRAME-INTERNAL ARITHMETIC (per-prime structure)
     p | D_(p-1)==!p | 

In [6]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:D4 (p13020) [T] — the deciding family e.R2
# Horizon determination of the readout: $\ro(\varepsilon_n)=\ro(\eR)$ for $n\ge n_0(k)=\Theta(k/\log k)$; the
# binary64 constant for $e$ is the readout of $\varepsilon_{18}=18!/\dr{18}=6402373705728000/2355301661033953$
# and $n=18$ is minimal ($\varepsilon_{17}$ fails readout equality,
# $|\varepsilon_{17}-\eR|=1.10\times10^{-15}$ above the half-ulp); the first $100$ significant digits of $\eR$
# are those of $\varepsilon_{70}$; every floating-point value of $e$ is a plateau readout.
from frc_13_epi import predicate; predicate("13:D4")

— epi.py:324 (the check that decides 13:D4; family e.R2)
  324      # 13:D4 (p13020), 13:I1 (p13048)
  325      family("e", "R2")
  326      chk("minimal n with float(n!/D_n) == binary64(e) is n = 18", n_double == 18)
  327      chk("eps_18 = 6402373705728000/2355301661033953", Fraction(fact[18], D[18]) == Fraction(6402373705728000, 2355301661033953))
  328      chk("eps_17 fails readout equality", float(Fraction(fact[17], D[17])) != target)
  329      chk("|eps_17 - e| exceeds the half-ulp 2^-53 and |eps_18 - e| lies below it", abs(Fraction(fact[17], D[17]) - E) > Fraction(1, 2**53) > abs(Fraction(fact[18], D[18]) - E))
  330  
  331      # (c) 100-digit determination
  332      n100 = None
  333      for n in range(2, NMAX):
  334          if Fraction(8, fact[n+1]) < Fraction(1, 10**100):
  335              n100 = n; break
  336      d_chain = frac_to_dec(Fraction(fact[n100], D[n100]), 130)
  337      d_e     = frac_to_dec(E, 130)
  [PASS] e.R2     (deciding)       4 exact micro-chec

In [7]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:D5 (p13021) [T] — the deciding family e.R4
# Feasibility selection: the compound chain $(1+1/n)^{n}$ costs $n\asymp2^{k}$ steps for $k$ bits against
# $n_0(k)=\Theta(k/\log k)$ for the derangement chain ($n=11,18,70$ at $k=24,53,333$); within a horizon $H$
# the compound chain realises $O(\log H)$ bits and the factorial chain $\Theta(H\log H)$; under the
# feasibility principle the derangement representation is the canonical finite definition of $e$.
from frc_13_epi import predicate; predicate("13:D5")

— epi.py:352 (the check that decides 13:D5; family e.R4)
  352      # 13:D5 (p13021)
  353      family("e", "R4")
  354      for k in (24, 53, 333):
  355          # derangement chain: 8/(n+1)! < 2^-k
  356          n1 = 2
  357          while Fraction(8, fact[n1+1]) >= Fraction(1, 2**k): n1 += 1
  358          # compound chain: |(1+1/n)^n - e| ~ e/(2n) -> n ~ e*2^{k-1}
  359          n2 = math.e * 2**(k-1)
  360          log_e(f"    k={k:3d}:  derangement chain n = {n1:3d}   |   (1+1/n)^n chain n ~ {n2:.2e}")
  361          chk(f"derangement chain reaches 2^-{k} at n = {(24, 53, 333).index(k) and [11, 18, 70][(24, 53, 333).index(k)] or 11}", n1 == [11, 18, 70][(24, 53, 333).index(k)])
  362  
  363      # =========================================================
  364      log_e(); log_e("="*72); log_e("[2] FRAME-INTERNAL ARITHMETIC (per-prime structure)")
  365      log_e("="*72)
  [PASS] e.R4     (deciding)       3 exact micro-checks
13:D5: VERIFIED — 1 family record(s)


In [8]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:E2 (p13023) [T] — the deciding family pi.R1
# Two-sided monotone enclosure: $v_n$ strictly increasing, $w_n$ strictly decreasing, $v_n<\piR<w_n$, and the
# exact width identity $w_n-v_n=w_n/(2n+1)$, certifying $\approx\log_2n$ bits by nested $\fQ$-intervals; the
# identification of the common limit with $\piR$ is the Wallis import (A6); verified for $n\le300$ against
# Machin brackets of width below $10^{-420}$.
from frc_13_epi import predicate; predicate("13:E2")

— epi.py:626 (the check that decides 13:E2; family pi.R1)
  626      # 13:C5 (p13016), 13:E2 (p13023)
  627      family("pi", "R1")
  628      chk("v_n < pi < w_n for n = 1..300 (Machin bracket)", encl)
  629      chk("exact width identity w_n - v_n = w_n/(2n+1)", ident)
  630      chk("v_n strictly increasing, w_n strictly decreasing", mono)
  631      # p=13 showcase residue line (round-01: anchors the manuscript's displayed line)
  632      p13_line = [pow(16, n, 13) * pow(n * pow(math.comb(2*n, n), 2, 13) % 13, -1, 13) % 13 for n in range(1, 7)]
  633      log_pi(f"    p=13 residue line [w_n]_13, n=1..6: {p13_line} " + ("PASS" if p13_line == [4, 5, 10, 9, 6, 11] else "FAIL"))
  634      # 13:G8 (p13041)
  635      family("pi", "R2")
  636      chk("p=13: [w_n]_13 for n = 1..6 is 4, 5, 10, 9, 6, 11", p13_line == [4, 5, 10, 9, 6, 11])
  637  
  638      # Machin chain: M_N with both arctans truncated at N
  639      def machin(N):
  [PASS] pi.R1    (deciding)       3 exact micro-chec

In [9]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:E3 (p13024) [T] — the deciding family pi.R4
# The arcsin chain $s_n=3\sum_{k\le n}\Cb{k}/((2k+1)16^{k})\in\fQ$ with external completion
# $\piR=6\arcsin\frac12$ \chart{} and tail bound $|\piR-s_n|<4^{-n}/(2n+3)$ for $n\ge1$: geometric rate from
# the same combinatorial atoms as the Wallis pair.
from frc_13_epi import predicate; predicate("13:E3")

— epi.py:674 (the check that decides 13:E3; family pi.R4)
  674      # 13:E3 (p13024)
  675      family("pi", "R4")
  676      s_run = Fraction(0); tail_ok = True
  677      for n in range(0, 121):
  678          s_run += Fraction(comb(2*n, n), (2*n+1) * 16**n)
  679          if n >= 1 and not abs(PI - 3*s_run) < Fraction(1, 4**n * (2*n+3)): tail_ok = False
  680      chk("arcsin chain s_n = 3 sum C(2k,k)/((2k+1)16^k): |pi - s_n| < 4^-n/(2n+3) for n = 1..120", tail_ok)
  681  
  682      # feasibility table
  683      log_pi("(c) precision-cost (terms n for 2^-k):")
  684      for k in (24, 53, 333):
  685          nL = 2**k            # Leibniz/Wallis ~ error 1/n
  686          nM = math.ceil((k*math.log(2)/math.log(5) - 3)/2)   # Machin ~ 5^{-2N}
  687          nA = math.ceil(k/2)  # arcsin(1/2) ~ 4^{-n}
  [PASS] pi.R4    (deciding)       1 exact micro-checks
13:E3: VERIFIED — 1 family record(s)


In [10]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:E4 (p13025) [T] — the deciding family pi.R3
# Horizon determination for $\pi$: the Machin chain $M_N\in\fQ$ has $|M_N-\piR|<17/((2N+3)5^{2N+3})$; the
# binary64 constant $884279719003555/281474976710656$ is the readout of $M_{10}$ with $N=10$ minimal; the
# first $100$ decimal digits are fixed by $M_{71}$; alternating pairing of the two arctangent series gives
# certified two-sided $\fQ$-enclosures at every truncation, the Gregory brackets $4S_{2m+1}<\piR<4S_{2m}$ of
# width $4/(4m+3)$ the base case.
from frc_13_epi import predicate; predicate("13:E4")

— epi.py:665 (the check that decides 13:E4; family pi.R3)
  665      # 13:E4 (p13025), 13:E5 (p13026), 13:I2 (p13049)
  666      family("pi", "R3")
  667      chk("minimal N with float(M_N) == binary64(pi) is N = 10", n_double == 10)
  668      chk("binary64(pi) = 884279719003555/281474976710656 = float(M_10)", Fraction(math.pi) == Fraction(884279719003555, 281474976710656) and float(machin(10)) == math.pi)
  669      chk("|M_N - pi| < 17/((2N+3) 5^(2N+3)) for N = 2..39", err_bound_ok)
  670      chk("100 decimal digits fixed at N = 71 (|M_71 - pi| < 10^-101, first N with that property from 60)", N100 == 71)
  671      chk("certified enclosure at (12, 4) terms contains pi and has width below 2^-52", lo < PI < hi and hi - lo < Fraction(1, 2**52))
  672      chk("Gregory brackets 4 S_(2m+1) < pi < 4 S_(2m) of width 4/(4m+3), m <= 40", all(4*Sg(2*m+1) < PI < 4*Sg(2*m) and 4*Sg(2*m) - 4*Sg(2*m+1) == Fraction(4, 4*m+3) for m in range(1, 41)))
  673      # arcsin chain tail bound (Definition

In [11]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:E5 (p13026) [T] — the deciding family pi.R3
# The feasibility hierarchy: among $\fQ$-native constructors of $\pi$ the best known rate is geometric,
# $\Theta(k)$ steps (Machin, arcsin), against $\Theta(k/\log k)$ for the derangement chain of $e$; the circle
# count and the Leibniz/Wallis chains sit at $2^{k}$; the AGM is not $\fQ$-native. A factorial-rate chain
# exists for every real by rounding (and for $\pi$ by rounding the Machin chain), so the split is by kind of
# chain --- E-partial sums ($e$) against G-partial sums ($\pi$), master C26; whether $\piR$ is the value of an
# E-function with rational coefficients at a rational point is a question of the continuum (Fischler--Rivoal's
# conjecture $\operatorname{Frac}\mathcal E\cap\operatorname{Frac}\mathcal G=\overline{\Q}$ would answer it
# negatively), not a predicate of the frame.
from frc_13_epi import predicate; predicate("13:E5")

— epi.py:665 (the check that decides 13:E5; family pi.R3)
  665      # 13:E4 (p13025), 13:E5 (p13026), 13:I2 (p13049)
  666      family("pi", "R3")
  667      chk("minimal N with float(M_N) == binary64(pi) is N = 10", n_double == 10)
  668      chk("binary64(pi) = 884279719003555/281474976710656 = float(M_10)", Fraction(math.pi) == Fraction(884279719003555, 281474976710656) and float(machin(10)) == math.pi)
  669      chk("|M_N - pi| < 17/((2N+3) 5^(2N+3)) for N = 2..39", err_bound_ok)
  670      chk("100 decimal digits fixed at N = 71 (|M_71 - pi| < 10^-101, first N with that property from 60)", N100 == 71)
  671      chk("certified enclosure at (12, 4) terms contains pi and has width below 2^-52", lo < PI < hi and hi - lo < Fraction(1, 2**52))
  672      chk("Gregory brackets 4 S_(2m+1) < pi < 4 S_(2m) of width 4/(4m+3), m <= 40", all(4*Sg(2*m+1) < PI < 4*Sg(2*m) and 4*Sg(2*m) - 4*Sg(2*m+1) == Fraction(4, 4*m+3) for m in range(1, 41)))
  673      # arcsin chain tail bound (Definition

In [12]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F1 (p13027) [T] — the deciding family e.W1
# Wilson reflection: for an odd prime $\p$ and $0\le k\le\p-1$, $1/k!\equiv-(-1)^{k}(\p-1-k)!\pmod\p$.
from frc_13_epi import predicate; predicate("13:F1")

— epi.py:400 (the check that decides 13:F1; family e.W1)
  400          # 13:F1 (p13027), 13:F2 (p13028)
  401          family("e", "W1")
  402          f_ = [1]
  403          for k in range(1, p): f_.append(f_[-1] * k % p)
  404          chk(f"p={p}: Wilson reflection 1/k! == -(-1)^k (p-1-k)! for all k < p", all(pow(f_[k], -1, p) == (-(-1)**k * f_[p-1-k]) % p for k in range(p)))
  405          chk(f"p={p}: wall identity D_(p-1) == K(p)", wall)
  406          chk(f"p={p}: K(p) != 0, the terminal residue -(K(p))^-1 exists", kp != 0)
  407          # 13:F3 (p13029)
  408          family("e", "W2")
  409          chk(f"p={p}: antiperiodicity D_(n+p) == -D_n over a full period", anti)
  410          chk(f"p={p}: 1 in Z0(p) and p-1 not in Z0(p)", 1 in z0 and (p-1) not in z0)
  411          # the tail identity and the collision reading of the blind set (13:F8; the Y1 push of 2026-09-18)
  412          # 13:F8 (p13066)
  413          family("e", "W5")
  [PASS] e.W1     (deciding)       21 ex

In [13]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F2 (p13028) [T] — the deciding family e.W1
# The wall identity and the terminal residue of $e$: $\dr{(\p-1)}\equiv\Ku{\p}\pmod\p$ (the odd-prime case of
# the Mijajlovi\'c--\v Sami congruence, A4), hence $\rep{\varepsilon_{\p-1}}=-(\Ku{\p})^{-1}$ whenever
# $\p\nmid\Ku{\p}$; existence of the terminal link on a shell is equivalent to $\p\nmid\Ku{\p}$, universal
# existence to Kurepa's hypothesis (Y1); re-verified in the derangement form for all $121{,}126$ odd primes
# $\p<1.6\times10^{6}$ with zero failures of either condition. Beyond the wall $n!\equiv0$: $n=\p-1$ is the
# factorial horizon.
from frc_13_epi import predicate; predicate("13:F2")

    primes_checked=22043 kurepa_vanishing=0 wall_congruence_failures=0
  [PASS] kur.K1  EXACT [13:F2, 13:J6, 13:Y1] Kurepa's wall in the derangement form for all 22 043 odd primes p < 2.5·10⁵ (kurepa_wall.c, one O(p) pass per   --  3 exact micro-checks; kurepa_wall.c, N = 250000, 29 s
— epi.py:400 (the check that decides 13:F2; family e.W1)
  400          # 13:F1 (p13027), 13:F2 (p13028)
  401          family("e", "W1")
  402          f_ = [1]
  403          for k in range(1, p): f_.append(f_[-1] * k % p)
  404          chk(f"p={p}: Wilson reflection 1/k! == -(-1)^k (p-1-k)! for all k < p", all(pow(f_[k], -1, p) == (-(-1)**k * f_[p-1-k]) % p for k in range(p)))
  405          chk(f"p={p}: wall identity D_(p-1) == K(p)", wall)
  406          chk(f"p={p}: K(p) != 0, the terminal residue -(K(p))^-1 exists", kp != 0)
  407          # 13:F3 (p13029)
  408          family("e", "W2")
  409          chk(f"p={p}: antiperiodicity D_(n+p) == -D_n over a full period", anti)
  410          chk(f"p=

In [14]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F3 (p13029) [T] — the deciding family e.W2
# Antiperiodicity: $\dr{(n+\p)}\equiv-\dr{n}\pmod\p$ for every odd prime and every $n\ge0$, so $\dr{n}\bmod\p$
# depends on $n\bmod2\p$; the blind set $Z_0(\p)=\{n<\p:\dr{n}\equiv0\}$ always contains $1$, and Kurepa's
# hypothesis is $\p-1\notin Z_0(\p)$ for every $\p$.
from frc_13_epi import predicate; predicate("13:F3")

— epi.py:407 (the check that decides 13:F3; family e.W2)
  407          # 13:F3 (p13029)
  408          family("e", "W2")
  409          chk(f"p={p}: antiperiodicity D_(n+p) == -D_n over a full period", anti)
  410          chk(f"p={p}: 1 in Z0(p) and p-1 not in Z0(p)", 1 in z0 and (p-1) not in z0)
  411          # the tail identity and the collision reading of the blind set (13:F8; the Y1 push of 2026-09-18)
  412          # 13:F8 (p13066)
  413          family("e", "W5")
  414          Kj = [0]
  415          for k in range(p): Kj.append((Kj[-1] + f_[k]) % p)          # Kj[j] = K(j) = sum_{k<j} k! mod p
  416          chk(f"p={p}: tail identity j! D_(p-1-j) == (-1)^j (K(p) - K(j)) for all j < p", all(f_[j] * d[p-1-j] % p == ((-1)**j * (Kj[p] - Kj[j])) % p for j in range(p)))
  417          chk(f"p={p}: blind set = collisions of the partial sums: Z0(p) = {{p-1-j : K(j) == K(p), j <= p-2}}", set(z0) == {p-1-j for j in range(p-1) if Kj[j] == Kj[p]})
  418          chk(f"p={p}: the force

In [15]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F4 (p13030) [T] — the deciding family e.W3
# Series duals at the wall: $\sum_{k<\p}(-1)^{k}/k!\equiv-\Ku{\p}$ and $\sum_{k<\p}1/k!\equiv-A(\p)$, $A(\p)$
# the alternating factorial sum; the group law $\exp(1)\exp(-1)=1$ does not survive,
# $\Ku{\p}A(\p)\equiv3,0,21,28,132,258$ at $\p=7,13,29,101,257,1009$; on $\p=13$ the direct series residue
# vanishes ($A(13)\equiv0$) while the derangement side survives ($\Ku{13}\equiv10$) --- a second ground for
# the canonicity of D5.
from frc_13_epi import predicate; predicate("13:F4")

— epi.py:438 (the check that decides 13:F4; family e.W3)
  438      # 13:F4 (p13030)
  439      family("e", "W3")
  440      chk("series duals S- == -K(p), S+ == -A(p) for p in [7, 13, 29, 101, 257, 1009]", dual_ok)
  441      chk("K(p) A(p) mod p = 3, 0, 21, 28, 132, 258 (never 1)", [v for _, v in prod_examples] == [3, 0, 21, 28, 132, 258])
  442      chk("asymmetry at p = 13: A(13) == 0 and K(13) == 10", alt_fact_mod(13) == 0 and kurepa_mod(13) == 10)
  443      # the p = 13 picture: the residue line of eps_n, n = 2..12
  444      # 13:F6 (p13032)
  445      family("e", "W4")
  446      d13 = d_mod(13, 13)
  447      line13 = [None if d13[n] == 0 else (math.factorial(n) % 13) * pow(d13[n], -1, 13) % 13 for n in range(2, 13)]
  448      log_e(f"  p=13 residue line [eps_n]_13, n=2..12: {['blind' if v is None else v for v in line13]}")
  449      chk("p=13: [eps_n]_13 for n = 2..12 is 2, 3, 7, 11, 1, 6, blind, 2, 8, 10, 9", line13 == [2, 3, 7, 11, 1, 6, None, 2, 8, 10, 9])
  450      ch

In [16]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F5 (p13031) [T] — the deciding family e.S1
# Blind-scale statistics: over the $501$ primes in $(1000,5000)$ the sizes $|Z_0(\p)|$ distribute as
# $1^{(169)},2^{(209)},3^{(92)},4^{(22)},5^{(6)},6^{(2)},7^{(1)}$, mean $1.996$, the excess over the forced
# $n=1$ matching Poisson$(1)$ \apx{} (the model: $\dr{n}\bmod\p$ uniform for $2\le n\le\p-2$, the wall
# excluded; extended to the $1593$ primes in $(5000,20000)$:
# $1^{(597)},2^{(569)},3^{(299)},4^{(108)},5^{(17)},6^{(3)}$, mean $1.988$); blind scales never obstruct the
# plateau of D4 (advance at most $|Z_0(\p)|$ steps past $n_0$). The counts are exact; the Poisson reading is
# the model they match.
from frc_13_epi import predicate; predicate("13:F5")

— epi.py:467 (the check that decides 13:F5; family e.S1)
  467      # 13:F5 (p13031)
  468      family("e", "S1")
  469      chk("501 primes in (1000, 5000)", len(primes_stat) == 501)
  470      chk("|Z0(p)| distribution 1:169, 2:209, 3:92, 4:22, 5:6, 6:2, 7:1", dict(cnt) == {1: 169, 2: 209, 3: 92, 4: 22, 5: 6, 6: 2, 7: 1})
  471      chk("mean |Z0(p)| = 1000/501 = 1.996", Fraction(sum(sizes), len(sizes)) == Fraction(1000, 501))
  472  
  473      # =========================================================
  474      log_e(); log_e("="*72); log_e("[3] ANGULAR UNIT e_t = g^{i_t}: radian calibration over p < 10^6")
  475      log_e("="*72)
  476      LIM = 1000000
  477      primes = sieve_e(LIM)
  478      PI_LO, PI_HI = pi_bracket(80)                 # certified rational bracket of pi, width < 1e-100
  479      best = []
  480      count_001 = 0
  [PASS] e.S1     (deciding)       3 exact micro-checks
13:F5: VERIFIED — 1 family record(s)


In [17]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F6 (p13032) [T] — the deciding family e.W4
# The $\p=13$ picture: the residue line $\rep{\varepsilon_n}$, $n=2,\dots,12$, is
# $2,3,7,11,1,6,\text{blind},2,8,10,9$ --- blind at $n=8$ ($\dr{8}=14833=13\cdot1141$) and terminating on
# $\rep{\varepsilon_{12}}=-(\Ku{13})^{-1}=9$, the common reading of $6/5$ and $11/7$ --- while the metric row
# freezes onto $2.71828\ldots$ by $n\approx9$: the thesis in miniature.
from frc_13_epi import predicate; predicate("13:F6")

— epi.py:444 (the check that decides 13:F6; family e.W4)
  444      # 13:F6 (p13032)
  445      family("e", "W4")
  446      d13 = d_mod(13, 13)
  447      line13 = [None if d13[n] == 0 else (math.factorial(n) % 13) * pow(d13[n], -1, 13) % 13 for n in range(2, 13)]
  448      log_e(f"  p=13 residue line [eps_n]_13, n=2..12: {['blind' if v is None else v for v in line13]}")
  449      chk("p=13: [eps_n]_13 for n = 2..12 is 2, 3, 7, 11, 1, 6, blind, 2, 8, 10, 9", line13 == [2, 3, 7, 11, 1, 6, None, 2, 8, 10, 9])
  450      chk("p=13: D_8 = 14833 = 13 * 1141 (the blind scale)", D[8] == 14833 == 13 * 1141)
  451      chk("p=13: the wall invariant 9 = -(K(13))^-1 = [6/5] = [11/7]", (-pow(kurepa_mod(13), -1, 13)) % 13 == 9 == 6 * pow(5, -1, 13) % 13 == 11 * pow(7, -1, 13) % 13)
  452  
  453      # blind-scale statistics
  454      log_e()
  455      primes_stat = [p for p in sieve_e(5000) if 1000 < p < 5000]
  456      sizes = []
  457      for p in primes_stat:
  [PASS] e.W4     (deciding)

In [18]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F7 (p13033) [T] — the deciding family tow.E
# The fixed-shell tower of $e$: with $\delta_m=\ctr{(-1)^{m}\eP}$ and $q_m=((m\p)!+\delta_m)/\dr{(m\p)}$,
# $\rep{q_m}=\eP$ exactly at every grade (since $(m\p)!\equiv0$ and $\dr{(m\p)}\equiv(-1)^{m}$ by F3) while
# $q_m\to\eR$ with the certificate $|q_m-\eR|<(m\p)!/(\dr{(m\p)}\dr{(m\p+1)})+2\kap/\dr{(m\p)}$; on $\p=13$,
# $q_2$ differs from $\eR$ by $3.98\times10^{-26}$ and every member reads $6$. Compatibility, not selection.
from frc_13_epi import predicate; predicate("13:F7")

— epi.py:1118 (the check that decides 13:F7; family tow.E)
 1118      # 13:B3 (p13010), 13:F7 (p13033)
 1119      family("tow", "E")
 1120      for m in (1, 2, 3):
 1121          delta = centered(((-1)**m * eP) % p, p)
 1122          num = math.factorial(m*p) + delta
 1123          den = subfact(m*p)
 1124          check_tow(f"p={p} e-tower grade {m}: !(mp) == (-1)^m (mod p)", den % p == ((-1)**m) % p)
 1125          check_tow(f"p={p} e-tower grade {m}: shell reading = e_p", (num * pow(den % p, p-2, p)) % p == eP)
 1126          lo, hi = e_brackets(max(m*p//2, 12))
 1127          q = F(num, den)
 1128          # the paper's certificate: |q - e| < (mp)!/( !(mp) !(mp+1) ) + pi_F/!(mp)
 1129          width = F(math.factorial(m*p), den*subfact(m*p+1)) + F((p-1)//2, den)
 1130          check_tow(f"p={p} e-tower grade {m}: |q - e| inside certified window",
 1131                q > lo - width and q < hi + width)
  [PASS] tow.E    (deciding)       18 exact micro-checks
13:F7: VERIFIED — 1 fami

In [19]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:F8 (p13066) [T] — the deciding family e.W5
# The tail identity: $j!\,\dr{(\p-1-j)}\equiv(-1)^{j}\bigl(\Ku{\p}-\Ku{j}\bigr)\pmod\p$ for $0\le j\le\p-1$
# ($j=0$ is F2); hence $n\in Z_0(\p)$ iff $\Ku{(\p-1-n)}\equiv\Ku{\p}$: the blind set counts the collisions of
# the partial sums of the left factorial with their total, $n=1$ is the forced collision
# $\Ku{(\p-2)}\equiv\Ku{\p}$, and Kurepa's hypothesis is the return-to-origin question of the walk with steps
# $j!$.
from frc_13_epi import predicate; predicate("13:F8")

— epi.py:412 (the check that decides 13:F8; family e.W5)
  412          # 13:F8 (p13066)
  413          family("e", "W5")
  414          Kj = [0]
  415          for k in range(p): Kj.append((Kj[-1] + f_[k]) % p)          # Kj[j] = K(j) = sum_{k<j} k! mod p
  416          chk(f"p={p}: tail identity j! D_(p-1-j) == (-1)^j (K(p) - K(j)) for all j < p", all(f_[j] * d[p-1-j] % p == ((-1)**j * (Kj[p] - Kj[j])) % p for j in range(p)))
  417          chk(f"p={p}: blind set = collisions of the partial sums: Z0(p) = {{p-1-j : K(j) == K(p), j <= p-2}}", set(z0) == {p-1-j for j in range(p-1) if Kj[j] == Kj[p]})
  418          chk(f"p={p}: the forced collision K(p-2) == K(p) (n = 1) and K(0) = 0 != K(p) (Kurepa at p)", Kj[p-2] == Kj[p] and Kj[p] != 0)
  419          log_e(f"{p:6d} |   {'PASS' if wall else 'FAIL'}      |          {'PASS' if anti else 'FAIL'}           |    {ke!s:>8}                | {z0 if len(z0)<=8 else str(z0[:8])+'...'} (|Z0|={len(z0)})")
  420  
  421      # series duals
  422 

In [20]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G1 (p13034) [T] — the deciding family pi.W1
# The legibility window: with $m=(\p-1)/2$, $\p\nmid\Cb{n}$ for $1\le n\le m$ and $\p\mid\Cb{n}$ for $m<n<\p$
# (Kummer, A3): the residue line of the Wallis chain is defined on exactly $[1,m]$, blind-free below the wall
# and totally blind on $(m,\p)$; on a frame shell the last legible scale $n=2\kap$ is the angular address of
# $-1$.
from frc_13_epi import predicate; predicate("13:G1")

— epi.py:718 (the check that decides 13:G1; family pi.W1)
  718          # 13:G1 (p13034), 13:G2 (p13035)
  719          family("pi", "W1")
  720          chk(f"p={p}: p does not divide C(2n,n) for 1 <= n <= m", legible)
  721          chk(f"p={p}: p divides C(2n,n) for m < n < p", blind)
  722          chk(f"p={p}: [w_m] == -2 == m^-1 == 4 pi_A", wall == p - 2 == pow(t2, -1, p) == 4 * t2 % p)
  723          chk(f"p={p}: the lower chain is blind exactly at the wall, 2m+1 = p", 2*t2 + 1 == p)
  724  
  725      # also p == 3 mod 4
  726      p3ok = True
  727      for p in [7, 23, 1031]:
  728          m = (p-1)//2
  729          C = 1
  730          for n in range(m): C = C * 2*(2*n+1) % p * pow(n+1, -1, p) % p
  731          wall = pow(16, m, p) * pow(m, -1, p) * pow(C*C % p, -1, p) % p
  [PASS] pi.W1    (deciding)       28 exact micro-checks
13:G1: VERIFIED — 1 family record(s)


In [21]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G2 (p13035) [T] — the deciding family pi.W1
# The half-wall terminus, the calibration face: $\rep{w_m}=m^{-1}=\piA^{-1}=4\piA\equiv-2$ for every odd
# prime, the return statistics trivialising at the wall ($16^{m}\equiv1$, $\Cb{m}^{2}\equiv1$) and exposing
# the naked inverse scale; the identities are the calibration web of $2\piA\equiv-1$, and
# $\rep{w_{\piA}}\piA\equiv1\iff(4\kap)^{2}\equiv1$; the lower chain is blind at the wall and nowhere below.
from frc_13_epi import predicate; predicate("13:G2")

— epi.py:718 (the check that decides 13:G2; family pi.W1)
  718          # 13:G1 (p13034), 13:G2 (p13035)
  719          family("pi", "W1")
  720          chk(f"p={p}: p does not divide C(2n,n) for 1 <= n <= m", legible)
  721          chk(f"p={p}: p divides C(2n,n) for m < n < p", blind)
  722          chk(f"p={p}: [w_m] == -2 == m^-1 == 4 pi_A", wall == p - 2 == pow(t2, -1, p) == 4 * t2 % p)
  723          chk(f"p={p}: the lower chain is blind exactly at the wall, 2m+1 = p", 2*t2 + 1 == p)
  724  
  725      # also p == 3 mod 4
  726      p3ok = True
  727      for p in [7, 23, 1031]:
  728          m = (p-1)//2
  729          C = 1
  730          for n in range(m): C = C * 2*(2*n+1) % p * pow(n+1, -1, p) % p
  731          wall = pow(16, m, p) * pow(m, -1, p) * pow(C*C % p, -1, p) % p
  [PASS] tow.W    (corroborating)  22 exact micro-checks
  [PASS] pi.W1    (deciding)       28 exact micro-checks
  [PASS] pi.W2    (corroborating)  1 exact micro-checks
13:G2: VERIFIED — 3 family reco

In [22]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G3 (p13036) [T] — the deciding family pi2.A1
# The quarter-wall two-squares invariant: for $\p\equiv1\pmod4$, $\p=a^{2}+b^{2}$ with $a$ odd,
# $a\equiv1\pmod4$, at the quarter scale $n=\kap$ (the address of $\im$)
# $\rep{w_\kap}=\kap^{-1}(2a)^{-2}=-(a^{2})^{-1}=(b^{2})^{-1}$, by Gauss's congruence (A3) with
# $\kap\equiv-4^{-1}$; the invariant exists precisely under the frame condition; verified for all $211$ primes
# $\p\equiv1\pmod4$ below $3000$.
from frc_13_epi import predicate; predicate("13:G3")

[A] QUARTER-WALL TWO-SQUARES INVARIANT, all p == 1 (mod 4), p < 3000
  Gauss congruence C(2t,t) == 2a* (mod p): PASS (211 primes)
  k(w_t) == -(a^2)^(-1) == (b^2)^(-1) (mod p): PASS
  example p=13 = 3^2+2^2: k(w_3) = 10 = -(9)^(-1) = 10

[B] ARCSIN CHAIN sigma_p := sum_(k=0)^((p-3)/2) C(2k,k)/((2k+1)16^k)
  sigma == 0 (mod p^2) for all 5 <= p < 300: PASS
  third-order Bernoulli law sigma/p^2 == (-1)^((p+1)/2) B_(p-3)/36 (mod p), 5<=p<300: PASS (60 primes)  [mod-p^3 reading of Sun Conj. 5.1 + Wolstenholme refinement]
  p, sigma/p^2 mod p, law: [(5, 4, 4), (7, 3, 3), (11, 5, 5), (13, 6, 6), (17, 15, 15), (19, 2, 2), (23, 10, 10), (29, 21, 21)]

[C] PROOF INGREDIENTS
  C(2k,k) == (-4)^k binom(m,k) (mod p), k < m: PASS
  Lerch: H_((p-1)/2) == -2 q_p(2) (mod p): PASS
  L_m = sum binom(m,j)(-1)^j/(2j+1) = 4^m (m!)^2/(2m+1)!, m<=60: PASS
  key identity A_m + B_m = 2 L_m over Q, via formal-antiderivative bookkeeping, m<=60: PASS

[D] BLIND-RANGE AND REVIVAL REFINEMENTS
  blind-range sum == p*E

In [23]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G4 (p13037) [T] — the deciding family pi.S1
# Second order at the half wall: $w_m\equiv-2+2\p(q_\p(4)-1)\pmod{\p^{2}}$ with $q_\p(4)\equiv2q_\p(2)$, exact
# to $\p^{3}$ through Morley (A3); the universal $-2$ persists to $\p^{2}$ exactly for the $\pi$-Wieferich
# primes $4^{\p-1}\equiv1+\p\pmod{\p^{2}}$ (OEIS A355959, G10), which are exactly $5$ and $45827$ below
# $1.08\times10^{11}$ \cite{oeisA355959}, re-verified here to $2^{35}$ (\src{pi\_wieferich128};
# $45827\equiv3\pmod4$, so $\p=5$ is the only Subject shell among them; two below $10^{6}$ against the range
# frequency $\sum_{\p<10^{6}}1/\p\approx2.4$); the normalised quotients equidistribute empirically \apx.
from frc_13_epi import predicate; predicate("13:G4")

— epi.py:756 (the check that decides 13:G4; family pi.S1)
  756      # 13:G4 (p13037)
  757      family("pi", "S1")
  758      chk("Morley (-1)^m C(p-1,m) == 4^(p-1) (mod p^3) for p in {13, 29, 37, 41, 53, 101}", morley_ok)
  759      chk("w_m == -2 + 2p(q_p(4) - 1) (mod p^2) for the same primes", formula_ok)
  760      chk("q_p(4) == 2 q_p(2) (mod p) for the same primes", all((pow(4, p-1, p*p) - 1)//p % p == 2 * ((pow(2, p-1, p*p) - 1)//p) % p for p in [13, 29, 37, 41, 53, 101]))
  761  
  762      # pi-Wieferich search: 4^(p-1) == 1 + p (mod p^2)
  763      hits = []
  764      LIM = 1000000
  765      for p in sieve_pi(LIM):
  766          if pow(4, p-1, p*p) == (1 + p) % (p*p):
  767              hits.append(p)
  768      log_pi(f"pi-Wieferich primes (W == -2 mod p^2), p < 10^6: {hits if hits else 'none'}  "
  769          f"(heuristic expectation ~ sum 1/p ~ {sum(1/q for q in sieve_pi(LIM)):.2f} ... over the range)")
  [PASS] pi.S1    (deciding)       3 exact micro-checks
  [PASS]

In [24]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G5 (p13038) [T] — the deciding family pi.L1
# Lucas revivals and self-similarity: beyond the wall $n\ge\p$ is legible for $w$ precisely when every
# base-$\p$ digit of $n$ is at most $m$ and $\p\nmid n$ (on $\p=13$ the two-digit revivals number
# $(2\kap)^{2}=36$); on revival scales $\Cb{n}$ factorises digitwise (Lucas, A3); at the first revival
# $\rep{v_\p}\equiv8$ universally and $v_\p\equiv8+16\p(2q_\p(2)-1)\pmod{\p^{2}}$ by Wolstenholme.
from frc_13_epi import predicate; predicate("13:G5")

— epi.py:813 (the check that decides 13:G5; family pi.L1)
  813      # 13:G5 (p13038)
  814      family("pi", "L1")
  815      chk("p=13: C(26,13) == 2 == C(2,1) C(0,0) (mod 13) and [v_13] == 8", Cv == 2 and vres == 8)
  816      chk("p=13: C(28,14) == 4 == C(2,1)^2 (mod 13), Lucas digitwise", C28 == 4)
  817      chk("p=13: two-digit revival scales in [13, 169) number exactly (2 kappa)^2 = 36", cnt == 36)
  818      chk("[v_p] == 8 universally, p in {5, 7, 11, 13, 29, 37}", all(2 * pow(16, q, q) * pow((2*q+1) % q * pow(comb(2*q, q), 2, q) % q, -1, q) % q == 8 % q for q in [5, 7, 11, 13, 29, 37]))
  819  
  820      # =========================================================
  821      log_pi(); log_pi("="*72); log_pi("[5] ARCSIN(1/2) CHAIN WALL: T(p) and second order")
  822      log_pi("="*72)
  823      def T_mod(p):
  824          m = (p-1)//2
  825          s = 0; C = 1  # C(0,0)
  826          inv16 = pow(16, -1, p)
  [PASS] pi.L1    (deciding)       4 exact micro-checks
  [PASS]

In [25]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G6 (p13039) [T] — the deciding family pi.V1
# First-order wall vanishing of the arcsin chain: $\sigma_\p=\sum_{k<m}\Cb{k}/((2k+1)16^{k})\equiv0\pmod\p$
# for every prime $\p\ge5$, $\sigma_3=1$ the sole exception; proved entirely in finite arithmetic through the
# key identity $A+B=2L$ (coefficient bookkeeping of the formal antiderivative of $(1-s^{2})^{m}$), Wilson
# reflection, Lerch's congruence and the binomial transfer $\Cb{k}\equiv(-4)^{k}\binom mk$. Verified for all
# $428$ odd primes $5\le\p<3000$; the ingredients for $m\le60$, $\p<500$.
from frc_13_epi import predicate; predicate("13:G6")

— epi.py:842 (the check that decides 13:G6; family pi.V1)
  842      # 13:G6 (p13039)
  843      family("pi", "V1")
  844      chk("sigma_p == 0 (mod p) for all 428 odd primes 5 <= p < 3000", zero_all and tested == 428)
  845      chk("sigma_3 = 1, the sole exception", T_mod(3) == 1)
  846  
  847      # second order: sigma/p mod p, pattern hunt
  848      def euler_numbers_mod(p, upto):
  849          E = [0]*(upto+1); E[0] = 1
  850          for n in range(1, upto//2 + 1):
  851              s = 0
  852              for j in range(n):
  853                  s += comb(2*n, 2*j) * E[2*j]
  854              E[2*n] = (-s) % p
  855          return E
  [PASS] pi.V1    (deciding)       46 exact micro-checks
  [PASS] pi2.C1   (corroborating)  1 exact micro-checks
  [PASS] pi2.C2   (corroborating)  1 exact micro-checks
  [PASS] pi2.C3   (corroborating)  1 exact micro-checks
  [PASS] pi2.C4   (corroborating)  1 exact micro-checks
13:G6: VERIFIED — 5 family record(s)


In [26]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G8 (p13041) [T] — the deciding family pi.R2
# The $\p=13$ picture for $\pi$: $\kap=3$, wall at $n=6$; the residue line $\rep{w_n}$, $n=1,\dots,6$, is
# $4,5,10,9,6,11$ with $\rep{w_6}=11\equiv-2$ the calibration face and $\rep{w_3}=10\equiv-(3^{2})^{-1}$
# encoding $13=3^{2}+2^{2}$; empty blind set below the wall, total blindness for $n=7,\dots,12$, first revival
# $\rep{v_{13}}=8$.
from frc_13_epi import predicate; predicate("13:G8")

— epi.py:634 (the check that decides 13:G8; family pi.R2)
  634      # 13:G8 (p13041)
  635      family("pi", "R2")
  636      chk("p=13: [w_n]_13 for n = 1..6 is 4, 5, 10, 9, 6, 11", p13_line == [4, 5, 10, 9, 6, 11])
  637  
  638      # Machin chain: M_N with both arctans truncated at N
  639      def machin(N):
  640          def S(x, N): return sum(Fraction((-1)**k, (2*k+1)*x**(2*k+1)) for k in range(N+1))
  641          return 16*S(5, N) - 4*S(239, N)
  642      n_double = None
  643      for N in range(2, 25):
  644          if float(machin(N)) == math.pi:
  645              n_double = N; break
  646      log_pi(f"(b) Machin chain: minimal N with float(M_N) == IEEE-double(pi): N = {n_double}")
  647      err_bound_ok = all(abs(machin(N) - PI) < Fraction(17, (2*N+3)*5**(2*N+3)) for N in range(2, 40))
  [PASS] pi.R2    (deciding)       1 exact micro-checks
  [PASS] pi.L1    (corroborating)  4 exact micro-checks
13:G8: VERIFIED — 2 family record(s)


In [27]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G9 (p13042) [T] — the deciding family tow.P
# The fixed-shell tower of $\pi$: for $n=\p^{r}$, with $\delta_\p=\ctr{\piA B_n-A_n}=\ctr{-34}$, the member
# $\widehat\pi_{\p,r}=(2\cdot16^{\p^{r}}+\delta_\p)/((2\p^{r}+1)\Cb{\p^{r}}^{2})$ reads
# $\rep{\widehat\pi_{\p,r}}=\piA$ exactly for every $r$ (Lucas and Fermat give $\Cb{n}\equiv2$,
# $16^{n}\equiv16$, $B_n\equiv4$) while converging externally to $\piR$; on $\p=13$, $\delta_{13}=5$.
from frc_13_epi import predicate; predicate("13:G9")

— epi.py:1134 (the check that decides 13:G9; family tow.P)
 1134      # 13:G9 (p13042)
 1135      family("tow", "P")
 1136      for r in (1, 2):
 1137          n = p**r
 1138          Cb = C(2*n, n)
 1139          Bn = (2*n+1) * Cb * Cb
 1140          check_tow(f"p={p} pi-tower r={r}: C(2n,n) == 2 (mod p)", Cb % p == 2)
 1141          check_tow(f"p={p} pi-tower r={r}: 16^n == 16 (mod p)", pow(16, n, p) == 16 % p)
 1142          delta = centered((piA * Bn - 2*pow(16, n)) % p, p)
 1143          check_tow(f"p={p} pi-tower r={r}: delta = centered(-34)", delta == centered(-34 % p, p))
 1144          num = 2*pow(16, n) + delta
 1145          check_tow(f"p={p} pi-tower r={r}: shell reading = pi_A",
 1146                (num * pow(Bn % p, p-2, p)) % p == piA)
 1147          if r == 1:
  [PASS] tow.P    (deciding)       20 exact micro-checks
13:G9: VERIFIED — 1 family record(s)


In [28]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:G10 (p13067) [T] — the deciding family pi.S3
# The $\pi$-Wieferich condition in three forms: $4^{\p-1}\equiv1+\p\pmod{\p^{2}}$ iff $q_\p(4)\equiv1$ iff
# $2q_\p(2)\equiv1\pmod\p$, with $q_\p(4)=2q_\p(2)+\p\,q_\p(2)^{2}$ exactly; Eisenstein's
# $2q_\p(2)\equiv-H_{(\p-1)/2}\pmod\p$ ($H_m=\sum_{j\le m}1/j$) makes it $H_{(\p-1)/2}\equiv-1$, and
# $(\p+2)^{\p-1}\equiv1\pmod{\p^{2}}$ identifies it with OEIS A355959 \cite{oeisA355959}.
from frc_13_epi import predicate; predicate("13:G10")

— epi.py:774 (the check that decides 13:G10; family pi.S3)
  774      # 13:G10 (p13067)
  775      family("pi", "S3")
  776      S3 = [5, 13, 29, 37, 41, 53, 101, 45827]
  777      def fq(a, p): return (pow(a, p-1, p*p) - 1)//p             # the Fermat quotient q_p(a) reduced to [0, p)
  778      def fq_exact(a, p): return (a**(p-1) - 1)//p               # the Fermat quotient as the integer (a^(p-1) - 1)/p
  779      def H(m, p): return sum(pow(j, -1, p) for j in range(1, m+1)) % p
  780      chk("q_p(4) = 2 q_p(2) + p q_p(2)^2 exactly (as integers) for p in {5, 13, 29, 37, 41, 53, 101, 45827}",
  781          all(fq_exact(4, p) == 2*fq_exact(2, p) + p*fq_exact(2, p)**2 for p in S3))
  782      chk("the three forms agree on the same primes: 4^(p-1) == 1 + p (mod p^2) iff q_p(4) == 1 iff 2 q_p(2) == 1 (mod p)",
  783          all((pow(4, p-1, p*p) == (1 + p) % (p*p)) == (fq(4, p) % p == 1) == (2*fq(2, p) % p == 1) for p in S3))
  784      chk("Eisenstein: 2 q_p(2) == -H_((p-1)/2) (mod p

In [29]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:H2 (p13044) [T] — the deciding family tow.W
# The exact angular carrier of $\pi$: $\chi(-1)=e^{\imR\piR}$ exactly in every shell, for every generator and
# either chirality, since $-1=\gen^{(\p-1)/2}$ sits at arc fraction one half --- Euler's identity as the
# half-turn tautology $\gen^{2\kap}=-1$ under the dictionary, with no calibration and no error term.
from frc_13_epi import predicate; predicate("13:H2")

— epi.py:1110 (the check that decides 13:H2; family tow.W)
 1110      # 13:H2 (p13044), 13:I4 (p13051), 13:J1 (p13052)
 1111      family("tow", "W")
 1112      check_tow(f"p={p}: i^2 = -1", (i*i) % p == p-1)
 1113      lam = i                              # label lift: canonical representative read as exponent
 1114      eP = pow(g, lam, p)
 1115      piA = (2*kap) % p
 1116  
 1117      # --- e-tower: q_m = ((mp)! + delta_m)/!(mp), shell reading = eP, external -> e ---
 1118      # 13:B3 (p13010), 13:F7 (p13033)
 1119      family("tow", "E")
 1120      for m in (1, 2, 3):
 1121          delta = centered(((-1)**m * eP) % p, p)
 1122          num = math.factorial(m*p) + delta
 1123          den = subfact(m*p)
  [PASS] tow.W    (deciding)       22 exact micro-checks
13:H2: VERIFIED — 1 family record(s)


In [30]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:H5 (p13047) [T] — the deciding family e.N1
# The null experiment: whether $\gen^{\lambda(\im)}\equiv-(\Ku{\p})^{-1}$ for some primitive root of either
# chirality is the index condition $\gcd(s,\p-1)=\gcd(\lambda(\pm\im),\p-1)$; over the $210$ primes
# $\p\equiv1\pmod4$, $5<\p<3000$, it is solvable for $94$ ($44.8\%$, against $96.0\pm6.6$ expected under
# independence), the correlation of the two angular addresses $-0.10$ at deviation $\approx0.07$ \apx: no
# invariant pointwise identification of the angular and radial carriers of $e$; the naive reading of $\eP$ as
# the number $2.71828\ldots$ is refuted.
from frc_13_epi import predicate; predicate("13:H5")

— epi.py:556 (the check that decides 13:H5; family e.N1)
  556      # 13:H5 (p13047)
  557      family("e", "N1")
  558      chk("210 primes p == 1 (mod 4), 5 < p < 3000, with K(p) != 0", total == 210)
  559      chk("the index condition is solvable for exactly 94 of them", solvable == 94)
  560      _DIAG[0] = f"Pearson correlation of the angular addresses {num/den:+.3f} (the paper's [approx] reading)"
  561  
  562  
  563      flush("e", order=["R1", "R2", "R3", "R4", "W1", "W2", "W3", "W4", "W5", "S1", "A1", "N1"], details={"N1": _DIAG[0]})
  564  
  565  # ------------------------------------------------------------------------------------------------------------
  566  # block pi — validate_pi.py: FRC emergence of pi: verification suite.
  567  # Sections:
  568  #   [1] Radial chains: Wallis central-binomial pair (v_n, w_n) enclosure; Machin readout.
  569  #   [2] Wall arithmetic: legibility window [1,2t], empty blind set, hard wall, residue -2.
  [PASS] e.N1     (deciding)    

In [31]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:I1 (p13048) [T] — the deciding family e.R2
# Emergence of the classical $e$: (i) in every shell with $\p>n_0+|Z_0(\p)|+1$ a non-blind scale in
# $[n_0,n_0+|Z_0(\p)|]$ reads out $\ro(\eR)$, at binary64 exactly $18!/\dr{18}$; (ii) the frame-internal
# continuation is the antiperiodic residue line with blind set $Z_0(\p)$ and terminal invariant
# $-(\Ku{\p})^{-1}$, collapsed by the readout to one dyadic rational; (iii) $\chi(\eP)\to e^{\imR}$ along
# calibrated shells, exact calibration impossible, no pointwise link between the carriers. Composed from D4,
# F2, F3, F5, H3--H5.
from frc_13_epi import predicate; predicate("13:I1")

— epi.py:324 (the check that decides 13:I1; family e.R2)
  324      # 13:D4 (p13020), 13:I1 (p13048)
  325      family("e", "R2")
  326      chk("minimal n with float(n!/D_n) == binary64(e) is n = 18", n_double == 18)
  327      chk("eps_18 = 6402373705728000/2355301661033953", Fraction(fact[18], D[18]) == Fraction(6402373705728000, 2355301661033953))
  328      chk("eps_17 fails readout equality", float(Fraction(fact[17], D[17])) != target)
  329      chk("|eps_17 - e| exceeds the half-ulp 2^-53 and |eps_18 - e| lies below it", abs(Fraction(fact[17], D[17]) - E) > Fraction(1, 2**53) > abs(Fraction(fact[18], D[18]) - E))
  330  
  331      # (c) 100-digit determination
  332      n100 = None
  333      for n in range(2, NMAX):
  334          if Fraction(8, fact[n+1]) < Fraction(1, 10**100):
  335              n100 = n; break
  336      d_chain = frac_to_dec(Fraction(fact[n100], D[n100]), 130)
  337      d_e     = frac_to_dec(E, 130)
  [PASS] e.R2     (deciding)       4 exact micro-chec

In [32]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:I2 (p13049) [T] — the deciding family pi.R3
# Emergence of the classical $\pi$: (i) the plateau readout of $\fQ$-native chains, at binary64 exactly
# $M_{10}$, the Wallis pair certifying two-sided enclosures at every scale; (ii) the residue line on
# $[1,2\kap]$ with the quarter-wall invariant $-(a^{2})^{-1}$, the half-wall terminus $-2$, the second-order
# Fermat quotient with $\pi$-Wieferich set $\{5,45827\}$, Morley exactness at $\p^{3}$, total blindness beyond
# the address of $-1$, Lucas revivals, the arcsin wall vanishing to $\p^{2}$ with the Bernoulli third order
# and the Euler content displaced beyond the wall; (iii) $\chi(-1)=e^{\imR\piR}$ exactly. Composed from E2,
# E4, G1--G7, H2.
from frc_13_epi import predicate; predicate("13:I2")

— epi.py:665 (the check that decides 13:I2; family pi.R3)
  665      # 13:E4 (p13025), 13:E5 (p13026), 13:I2 (p13049)
  666      family("pi", "R3")
  667      chk("minimal N with float(M_N) == binary64(pi) is N = 10", n_double == 10)
  668      chk("binary64(pi) = 884279719003555/281474976710656 = float(M_10)", Fraction(math.pi) == Fraction(884279719003555, 281474976710656) and float(machin(10)) == math.pi)
  669      chk("|M_N - pi| < 17/((2N+3) 5^(2N+3)) for N = 2..39", err_bound_ok)
  670      chk("100 decimal digits fixed at N = 71 (|M_71 - pi| < 10^-101, first N with that property from 60)", N100 == 71)
  671      chk("certified enclosure at (12, 4) terms contains pi and has width below 2^-52", lo < PI < hi and hi - lo < Fraction(1, 2**52))
  672      chk("Gregory brackets 4 S_(2m+1) < pi < 4 S_(2m) of width 4/(4m+3), m <= 40", all(4*Sg(2*m+1) < PI < 4*Sg(2*m) and 4*Sg(2*m) - 4*Sg(2*m+1) == Fraction(4, 4*m+3) for m in range(1, 41)))
  673      # arcsin chain tail bound (Definition

In [33]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:I4 (p13051) [T] — the deciding family tow.W
# CRT stability: under the composite lift to $q=\prod\p_i$ the half-wall invariant of $\pi$ glues to $-2$
# modulo every $q$, since $2\piA\equiv-1$ holds in every fibre, while the terminal content of $e$,
# $(-(\Ku{\p_i})^{-1})_i$, varies fibrewise. The two selectors resolve the $\p=13$ coincidence: the same
# residue $6$ reaches $\piR$ through C4 and $\eR$ through C1.
from frc_13_epi import predicate; predicate("13:I4")

— epi.py:1110 (the check that decides 13:I4; family tow.W)
 1110      # 13:H2 (p13044), 13:I4 (p13051), 13:J1 (p13052)
 1111      family("tow", "W")
 1112      check_tow(f"p={p}: i^2 = -1", (i*i) % p == p-1)
 1113      lam = i                              # label lift: canonical representative read as exponent
 1114      eP = pow(g, lam, p)
 1115      piA = (2*kap) % p
 1116  
 1117      # --- e-tower: q_m = ((mp)! + delta_m)/!(mp), shell reading = eP, external -> e ---
 1118      # 13:B3 (p13010), 13:F7 (p13033)
 1119      family("tow", "E")
 1120      for m in (1, 2, 3):
 1121          delta = centered(((-1)**m * eP) % p, p)
 1122          num = math.factorial(m*p) + delta
 1123          den = subfact(m*p)
  [PASS] tow.W    (deciding)       22 exact micro-checks
13:I4: VERIFIED — 1 family record(s)


In [34]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:I5 (p13068) [T] — the deciding family frm.I1
# The three frame invariants across shells: on the $4783$ shells $\p\equiv1\pmod4$ below $10^{5}$, the triple
# $(\Ku{\p}/\p,\ q_\p(4)/\p,\ \varphi/\piR)$ --- the wall of $e$ (F2), the wall of $\pi$ at second order (G4),
# the angular address $\varphi=\arg(a+b\imR)$ of $\p=a^{2}+b^{2}$, $a\equiv1\pmod4$ (G3) --- is exact on each
# shell ($\p=13$: $10,6,-3+2\imR$; $1093$: $q_\p(4)=0$, the Wieferich prime); means $0.502,0.499,0.499$ and
# variances $0.0828,0.0831,0.0832$ (uniform: $1/2$, $1/12$), correlations $-0.016,-0.008,+0.010$
# ($\sigma=0.0145$), $4\times4\times4$ chi-square $59.6$ on $63$ df, $a/\sqrt\p=\cos\varphi$
# arcsine-distributed (mean $0.004$, variance $0.500$) \apx: consistent with independence on the examined
# range.
from frc_13_epi import predicate; predicate("13:I5")

    4783 shells p = 1 (mod 4) below 100000; K/p: mean 0.5024 var 0.0828; q_p(4)/p: mean 0.4986 var 0.0831 (certified rational bounds, width 1e-12)
    phi/pi: mean 0.4987 var 0.0832 [approx]; a/sqrt p: mean 0.004 var 0.500 [approx] (arcsine law: 0, 1/2); uniform law: 1/2, 1/12 = 0.0833
    4x4x4 table against the product of the quarter bins: chi-square 59.58 on 63 df (exact rational 284959/4783) [approx reading: consistent with independence]
  [PASS] frm.I1  EXACT [13:I5, 13:Y3] the three frame invariants of the two constants on the 4 783 shells p ≡ 1 (mod 4) below 10⁵ (frame_invariants.  --  7 exact micro-checks; frame_invariants.c, N = 100000, 2 s
— epi.py:1415 (the check that decides 13:I5; family frm.I1)
 1415      # 13:I5 (p13068), 13:Y3 (p13063)
 1416      family("frm", "I1")
 1417      if bound == N_frm:
 1418          chk("exactly 4783 shells p = 1 (mod 4) below 10^5", n == 4783)
 1419          chk("the triples at p = 13, 233, 30089 are (10, 6, -3+2i), (69, 14, 13+8i), (5260, 7

In [35]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:J1 (p13052) [T] — the deciding family tow.W
# The wrap-free window: an entity of the shell is $\p$-hard when it lies beyond the observer's horizon
# $\sqrt\p$; the accessible smalls are the natural counts below $\sqrt\p$, where sums and products of two
# accessible counts do not wrap, so primality, parity and order are wrap-invariant there and only there.
from frc_13_epi import predicate; predicate("13:J1")

— epi.py:1110 (the check that decides 13:J1; family tow.W)
 1110      # 13:H2 (p13044), 13:I4 (p13051), 13:J1 (p13052)
 1111      family("tow", "W")
 1112      check_tow(f"p={p}: i^2 = -1", (i*i) % p == p-1)
 1113      lam = i                              # label lift: canonical representative read as exponent
 1114      eP = pow(g, lam, p)
 1115      piA = (2*kap) % p
 1116  
 1117      # --- e-tower: q_m = ((mp)! + delta_m)/!(mp), shell reading = eP, external -> e ---
 1118      # 13:B3 (p13010), 13:F7 (p13033)
 1119      family("tow", "E")
 1120      for m in (1, 2, 3):
 1121          delta = centered(((-1)**m * eP) % p, p)
 1122          num = math.factorial(m*p) + delta
 1123          den = subfact(m*p)
  [PASS] tow.W    (deciding)       22 exact micro-checks
13:J1: VERIFIED — 1 family record(s)


In [36]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:J2 (p13053) [T] — the deciding family tow.H
# Calibration pinning: the half-period satisfies the height-two relation $2\piA+1\equiv0$, the shell
# calibration $\p=4\kap+1$ itself, so the observer holds an exact short certificate of an entity it never
# accesses as a tally; the same pinning covers the quarter-turn ($\im^{2}+1\equiv0$) and the residue web of
# G2; the pin holds on every one of the $500$ shells $\p\le8009$ with $H(\piA)=2$.
from frc_13_epi import predicate; predicate("13:J2")

— epi.py:1211 (the check that decides 13:J2; family tow.H)
 1211    # 13:J2 (p13053), 13:J3 (p13054), 13:J5 (p13056)
 1212    family("tow", "H")
 1213    hs = []
 1214    pin_ok = True; band_ok = True
 1215    for p in [q for q in _sieve(8009) if q % 4 == 1]:
 1216        kap = (p-1)//4
 1217        g = primitive_root(p)                      # iterates from 2: the smallest primitive root
 1218        e_neg = pow(g, pow(g, p-1-kap, p), p)      # exponential unit, oriented chirality  i = g^{-kap}
 1219        e_pos = pow(g, pow(g, kap, p), p)          # exponential unit, conjugate chirality i' = g^{+kap}
 1220        piA = (2*kap) % p
 1221        if not ((piA * (-2)) % p == 1 and minheight(piA, p) == 2): pin_ok = False
 1222        h = min(minheight(e_neg, p), minheight(e_pos, p))
 1223        if h > 2*_m.isqrt(p): band_ok = False
 1224        hs.append((p, h, _m.isqrt(p)))
  [PASS] tow.H    (deciding)       4 exact micro-checks; median H/sqrt(p) = 0.525 (diagnostic)
  [PASS] tow.W    (

In [37]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:J3 (p13054) [T] — the deciding family tow.H
# The height run: with $H(x)=\min_{a\ge1}\max(a,|\ctr{ax}|)$ in the smallest-primitive-root frame, minimised
# over the two chiralities, over the $500$ shells $\p\equiv1\pmod4$, $\p\le8009$: $H(\eP)\le2\sqrt\p$ on every
# shell (the horizon band), median $H(\eP)=0.525\sqrt\p$ \apx, exactly $68$ shells with $H\le10$, and the pin
# value $2$ on exactly $\{13,1933,4177,5857\}$; for $\eP$ every candidate bounded-height relation fails in
# almost all shells (the generic-growth reading is A7).
from frc_13_epi import predicate; predicate("13:J3")

— epi.py:1211 (the check that decides 13:J3; family tow.H)
 1211    # 13:J2 (p13053), 13:J3 (p13054), 13:J5 (p13056)
 1212    family("tow", "H")
 1213    hs = []
 1214    pin_ok = True; band_ok = True
 1215    for p in [q for q in _sieve(8009) if q % 4 == 1]:
 1216        kap = (p-1)//4
 1217        g = primitive_root(p)                      # iterates from 2: the smallest primitive root
 1218        e_neg = pow(g, pow(g, p-1-kap, p), p)      # exponential unit, oriented chirality  i = g^{-kap}
 1219        e_pos = pow(g, pow(g, kap, p), p)          # exponential unit, conjugate chirality i' = g^{+kap}
 1220        piA = (2*kap) % p
 1221        if not ((piA * (-2)) % p == 1 and minheight(piA, p) == 2): pin_ok = False
 1222        h = min(minheight(e_neg, p), minheight(e_pos, p))
 1223        if h > 2*_m.isqrt(p): band_ok = False
 1224        hs.append((p, h, _m.isqrt(p)))
  [PASS] tow.H    (deciding)       4 exact micro-checks; median H/sqrt(p) = 0.525 (diagnostic)
13:J3: VERIFIED — 1

In [38]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:J5 (p13056) [T] — the deciding family tow.H
# Universality and the duality as corollary: $\p$-hardness is self-similar, every shell finding its own
# half-period and drive name beyond its own horizon, so the chart shadows agree across all observers; $\pi$'s
# seat is calibration-pinned with a height-two certificate and its arc role normalisation-fixed, $e$ has
# neither a pinned seat nor a cross-shell certificate and its role is generator-dependent --- structural
# against statistical is the certification split (I3). The sporadic small-height shells of J3 are the
# certificate-level face of I4.
from frc_13_epi import predicate; predicate("13:J5")

— epi.py:1211 (the check that decides 13:J5; family tow.H)
 1211    # 13:J2 (p13053), 13:J3 (p13054), 13:J5 (p13056)
 1212    family("tow", "H")
 1213    hs = []
 1214    pin_ok = True; band_ok = True
 1215    for p in [q for q in _sieve(8009) if q % 4 == 1]:
 1216        kap = (p-1)//4
 1217        g = primitive_root(p)                      # iterates from 2: the smallest primitive root
 1218        e_neg = pow(g, pow(g, p-1-kap, p), p)      # exponential unit, oriented chirality  i = g^{-kap}
 1219        e_pos = pow(g, pow(g, kap, p), p)          # exponential unit, conjugate chirality i' = g^{+kap}
 1220        piA = (2*kap) % p
 1221        if not ((piA * (-2)) % p == 1 and minheight(piA, p) == 2): pin_ok = False
 1222        h = min(minheight(e_neg, p), minheight(e_pos, p))
 1223        if h > 2*_m.isqrt(p): band_ok = False
 1224        hs.append((p, h, _m.isqrt(p)))
  [PASS] tow.H    (deciding)       4 exact micro-checks; median H/sqrt(p) = 0.525 (diagnostic)
13:J5: VERIFIED — 1

In [39]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:J6 (p13057) [T] — the deciding family kur.K1
# Two constants, two walls: both attach their full-resolution shell content to sparse-prime problems (Kurepa
# non-vanishing for $e$, a prescribed Fermat quotient of Wieferich type for $\pi$); for each shell the bounded
# observer holds an exact finite certificate and holds no uniform certificate across all shells ---
# per-instance decidability without a uniform bound, the characteristic shape of statements at the arithmetic
# horizon; Kurepa's hypothesis is not thereby decided; the frequency of the persistence shells is a range
# reading (G4), their totality not a predicate of the paper.
from frc_13_epi import predicate; predicate("13:J6")

— epi.py:1301 (the check that decides 13:J6; family kur.K1)
 1301          # 13:J6 (p13057), 13:Y1 (p13061)
 1302          family("kur", "K1")
 1303          chk(f"{checked} odd primes p < {N_kur}: exactly 22043", checked == 22043)
 1304          chk("K(p) != 0 (mod p) for every prime checked", kfail == 0)
 1305          chk("!(p-1) == K(p) (mod p) for every prime checked", cfail == 0)
 1306          flush("kur", details={"K1": f"kurepa_wall.c, N = {N_kur}, {time.time() - t:.0f} s"})
 1307      else:
 1308          n = 20000
 1309          checked, kfail, cfail = python_pass_kur(n)
 1310          family("kur", "K1")
 1311          chk(f"{checked} odd primes p < {n}: exactly 2261", checked == 2261)
 1312          chk("K(p) != 0 (mod p) for every prime checked", kfail == 0)
 1313          chk("!(p-1) == K(p) (mod p) for every prime checked", cfail == 0)
 1314          flush("kur", details={"K1": f"NO C COMPILER: pure-Python pass to N_kur = {n} only ({checked} primes), {time.time() - t:.0

In [40]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:Y1 (p13061) [O] — the deciding family kur.K1
# Kurepa's hypothesis, $\p\nmid\Ku{\p}$ for every odd prime, equivalently $\p\nmid\dr{(\p-1)}$ (F2) and
# $B_{\p-1}\not\equiv1\pmod\p$: universal existence of the terminal residue $-(\Ku{\p})^{-1}$ of $e$;
# equivalently, the walk $\Ku{0},\Ku{1},\dots,\Ku{\p}$ of the left factorial's partial sums does not return to
# its origin modulo $\p$ (F8). Open since 1971; the one claimed proof (Barsky--Benzaghou 2004, Thm.~3)
# withdrawn by its authors (2011); no counterexample for $\p<2^{40}$; re-verified here with the wall identity
# for all $121{,}126$ odd primes $\p<1.6\times10^{6}$. Decided by no shell (each decides its own value:
# $\Ku{13}\equiv10$, $\Ku{233}\equiv69$, $\Ku{30089}\equiv5260$), by no finite search (F5's uniform model at
# the wall expects $\sum_{\p\le X}1/\p\approx\ln\ln X-0.24$ counterexamples \apx, so the model is scoped to
# $n\le\p-2$) and by no relation of this paper. Beside it, two large-$\p$ laws read on ranges:
# $|Z_0(\p)|-1\to\mathrm{Poisson}(1)$ (F5) and the independence of the quarter-turn index and the gcd class of
# the wall-residue index (H5; $94$ of $210$ against $96.0\pm6.6$ \apx).
from frc_13_epi import predicate; predicate("13:Y1")

— epi.py:1301 (the check that decides 13:Y1; family kur.K1)
 1301          # 13:J6 (p13057), 13:Y1 (p13061)
 1302          family("kur", "K1")
 1303          chk(f"{checked} odd primes p < {N_kur}: exactly 22043", checked == 22043)
 1304          chk("K(p) != 0 (mod p) for every prime checked", kfail == 0)
 1305          chk("!(p-1) == K(p) (mod p) for every prime checked", cfail == 0)
 1306          flush("kur", details={"K1": f"kurepa_wall.c, N = {N_kur}, {time.time() - t:.0f} s"})
 1307      else:
 1308          n = 20000
 1309          checked, kfail, cfail = python_pass_kur(n)
 1310          family("kur", "K1")
 1311          chk(f"{checked} odd primes p < {n}: exactly 2261", checked == 2261)
 1312          chk("K(p) != 0 (mod p) for every prime checked", kfail == 0)
 1313          chk("!(p-1) == K(p) (mod p) for every prime checked", cfail == 0)
 1314          flush("kur", details={"K1": f"NO C COMPILER: pure-Python pass to N_kur = {n} only ({checked} primes), {time.time() - t:.0

In [41]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
# 13:Y3 (p13063) [O] — the deciding family frm.I1
# The large-$\p$ laws of the two walls, read on ranges and open as laws. (a) The calibration rate
# $\delta(X)=\min_{\p\le X}|\theta-1|$: exact to $X=4\times10^{6}$ (H4), $\delta\to0$ by A7; whether its order
# is the uniform model's $\piR/(2N(X))$ for every large finite $X$, the continuum bound being the
# irrationality measure of $\piR$ ($\mu\le7.1033$, Zeilberger--Zudilin 2020) \import. (b) The joint law of
# $(\Ku{\p}/\p,\ q_\p(4)/\p,\ \varphi/\piR)$ for large $\p$ (I5 on $\p<10^{5}$): $\varphi/\piR$
# equidistributes (Hecke) \import, $q_\p(4)/\p$ equidistributes in the base at fixed $\p$ (Heath-Brown) and
# nothing is proved in $\p$, $\Ku{\p}/\p$ is Y1's; whether the three are independent in the large-$\p$ law.
# Each clause is a statement about every large finite $\p$ with the continuum as its chart, decided by no
# shell alone and by no relation of this paper.
from frc_13_epi import predicate; predicate("13:Y3")

— epi.py:1415 (the check that decides 13:Y3; family frm.I1)
 1415      # 13:I5 (p13068), 13:Y3 (p13063)
 1416      family("frm", "I1")
 1417      if bound == N_frm:
 1418          chk("exactly 4783 shells p = 1 (mod 4) below 10^5", n == 4783)
 1419          chk("the triples at p = 13, 233, 30089 are (10, 6, -3+2i), (69, 14, 13+8i), (5260, 7666, -67+160i)",
 1420              byp.get(13) == (10, 6, -3, 2) and byp.get(233) == (69, 14, 13, 8) and byp.get(30089) == (5260, 7666, -67, 160))
 1421          chk("the Wieferich prime 1093 is the one shell with q_p(4) = 0 (q_p(2) = 0 gives q_p(4) = 2q_p(2) + p q_p(2)^2 = 0)",
 1422              byp.get(1093, (0, -1))[1] == 0 and sum(1 for p, K, q, a, b in rows if q == 0) == 1)
 1423          chk("means of K/p and q_p(4)/p round to 0.502 and 0.499, variances to 0.0828 and 0.0831 (certified rational bounds)",
 1424              rounds_to(meanK, 3, Fr(502, 1000)) and rounds_to(meanQ, 3, Fr(499, 1000)) and rounds_to(varK, 4, Fr(828, 10000)) and round

## Summary

The cells above are the paper's python-witnessed predicates; the whole script, family by family, is `python3 epi.py`
(41 family checks, 813 exact micro-checks, `results.json`), `python3 -m frc_13_epi` once installed.

In [42]:
!pip install -q frc-13-epi --find-links https://finitering.space/pkg/
from frc_13_epi import verify_all
assert verify_all(), "a family check failed"
print("all family checks pass")


SUMMARY: 41/41 checks passed (813 exact micro-checks)
all family checks pass
